python/amici/petab_objective.py

```diff
@@ -1,181 +1,534 @@
-"""Functionality related to evaluating the objective function for a PEtab
-problem"""
+"""Functionality related to running simulations or evaluating the objective
+function as defined by a PEtab problem"""
 
-import numbers
-import numpy as np
-import pandas as pd
 import copy
+import logging
+import numbers
+from typing import List, Sequence, Optional, Dict, Tuple, Union, Any
 
-from typing import List, Sequence
 import amici
+import numpy as np
+import pandas as pd
 import petab
+from amici.logging import get_logger
+from petab.C import *
+
+
+logger = get_logger(__name__)
+
+
+def simulate_petab(petab_problem: petab.Problem,
+                   amici_model: amici.Model,
+                   solver: Optional[amici.Solver] = None,
+                   problem_parameters: Optional[Dict[str, float]] = None,
+                   log_level: int = logging.WARNING
+                   ) -> Tuple[float, List[amici.ReturnData]]:
+    """Simulate PEtab model
+
+    Arguments:
+        petab_problem: PEtab problem to work on
+        amici_model: AMICI Model assumed to be compatible with
+            ``petab_problem``
+        solver: An AMICI solver. Will use default options if None.
+        problem_parameters: Run simulation with these parameters. If None,
+            PEtab `nominalValues` will be used). To be provided as dict,
+            mapping PEtab problem parameters to SBML IDs.
+        log_level: Log level, see `logging` module
+    Returns:
+        Tuple of cost function value and a list of `ReturnData`s, corresponding
+        to the different simulation conditions. For ordering of simulation
+        conditions, see
+        `petab.Problem.get_simulation_conditions_from_measurement_df`.
+    """
+    logger.setLevel(log_level)
+
+    if solver is None:
+        solver = amici_model.getSolver()
+
+    if problem_parameters is None:
+        # Use PEtab nominal values as default
+        problem_parameters = {t.Index: getattr(t, NOMINAL_VALUE) for t in
+                              petab_problem.parameter_df.itertuples()}
+
+    # Generate ExpData with all condition-specific information
+    edatas = edatas_from_petab(model=amici_model,
+                               petab_problem=petab_problem,
+                               problem_parameters=problem_parameters)
+
+    # Simulate
+    rdatas = amici.runAmiciSimulations(amici_model, solver, edata_list=edatas)
+
+    # Compute total llh
+    llh = sum(rdata['llh'] for rdata in rdatas)
+
+    # log results
+    sim_cond = petab_problem.get_simulation_conditions_from_measurement_df()
+
+    for i, rdata in enumerate(rdatas):
+        logger.debug(f"Condition: {sim_cond.iloc[i, :].values}, status: "
+                     f"{rdata['status']}, llh: {rdata['llh']}")
+
+    return llh, rdatas
 
 
 def edatas_from_petab(
```

Comment:
wow. had a short look over it, did not spot mistakes, so fine if it works ;)

Reply:
So far, so good ;)